examples/raytracing/example_raytracing.cpp

```diff
@@ -0,0 +1,300 @@
+
+  Kokkos::View<ArborX::Sphere *, MemorySpace> _spheres;
+  Kokkos::View<float *, MemorySpace> _accumulator;
+
+  KOKKOS_FUNCTION float dotProduct(Vector const &v1, Vector const &v2) const
+  {
+    return v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2];
+  }
+
+  // Solves a*x^2 + b*x + c = 0.
+  // If a solution exists, return true and stores roots at x1, x2.
+  // If a solution does not exist, returns false.
+  KOKKOS_FUNCTION bool solveQuadratic(float a, float b, float c, float &x1,
+                                      float &x2) const
+  {
+    assert(a != 0);
+
+    auto discriminant = b * b - 4 * a * c;
+    if (discriminant < 0)
+      return false;
+    else if (discriminant == 0)
+    {
+      x1 = x2 = -b / (2 * a);
+      return true;
+    }
+
+    // Instead of doing a simple
+    //   0.5 * (-b +- std::sqrt(discriminant)) / (2*a)
+    // we use a more stable algorithm with less loss of precision (see, for
+    // example, https://www.scratchapixel.com/lessons/3d-basic-rendering/
+    // minimal-ray-tracer-rendering-simple-shapes/ray-sphere-intersection).
+    auto sgn = [](float x) { return (x > 0) - (x < 0); };
+    auto q = 0.5 * (-b - sgn(b) * std::sqrt(discriminant));
+    x1 = q / a;
+    x2 = c / q;
+
+    return true;
+  }
+
+  // Ray-Sphere intersection algorithm.
+  //
+  // The sphere can be expressed as the solution to
+  //     |p - c|^2 - r^2 = 0,           (1)
+  // where c is the center of the sphere, and r is the radius. On the other
+  // hand, any point on a bidirectional ray satisfies
+  //     p = o + t*d,                   (2)
+  // where o is the origin, and d is the direction vector.
+  // Substituting (2) into (1),
+  //     |(o + t*d) - c|^2 - r^2 = 0,   (3)
+  // resulting in a quadratic equation for unknown t
+  //     a2 * t^2 + a1 * t + a0 = 0
+  // with
+  //     a2 = |D|^2, a1 = 2*(d, o - c), and a0 = |o - c|^2 - r^2.
+  // Then, we only need to intersect the solution internval [tmin, tmax] with
+  // [0, +inf) of unidirectional ray.
+  KOKKOS_FUNCTION float overlapDistance(ArborX::Ray const &ray,
+                                        ArborX::Sphere const &sphere) const
+  {
+    auto const &c = sphere.centroid();
+    auto const &r = sphere.radius();
+    auto const &o = ray.origin();
+    auto const &d = ray.direction();
+
+    // Vector oc = o - c
+    Vector oc{o[0] - c[0], o[1] - c[1], o[2] - c[2]};
+
+    float a2 = 1.f; // directions are normalized
+    float a1 = 2.0 * dotProduct(d, oc);
+    float a0 = dotProduct(oc, oc) - r * r;
+
+    float t1;
+    float t2;
+    if (!solveQuadratic(a2, a1, a0, t1, t2))
+    {
+      // No intersection of a bidirectional ray with the sphere
+      return 0.f;
+    }
+    float tmin = KokkosExt::min(t1, t2);
+    float tmax = KokkosExt::max(t1, t2);
+
+    if (tmax < 0)
+    {
+      // Half-ray does not intersect with the sphere
+      return 0.f;
+    }
+
+    // Overlap [tmin, tmax] with [0, +inf)
+    tmin = KokkosExt::max(0.f, tmin);
```

Comment:
```suggestion
```

Reply:
No, removing this is wrong. Why would you want to do it?

Reply:
I wanted to condense the initialization, see https://github.com/arborx/ArborX/pull/461#discussion_r567255945.

Reply:
Ah, I see what you mean. I still don't think it should be done. Conceptually, tmin and tmax are assigned to be the min and max roots of the quadratic equation. The logic dealing with uni-directionality of the ray should be separate from that.